app/web_conn.go

```diff
@@ -187,29 +216,28 @@ func (c *WebConn) WritePump() {
 func (webCon *WebConn) InvalidateCache() {
 	webCon.AllChannelMembers = nil
 	webCon.LastAllChannelMembersTime = 0
-	webCon.SessionExpiresAt = 0
-	webCon.Session = nil
+	webCon.SetSession(model.Session{})
+	webCon.SetSessionExpiresAt(0)
 }
 
 func (webCon *WebConn) IsAuthenticated() bool {
 	// Check the expiry to see if we need to check for a new session
-	if webCon.SessionExpiresAt < model.GetMillis() {
-		if webCon.SessionToken == "" {
+	if webCon.GetSessionExpiresAt() < model.GetMillis() {
+		if webCon.GetSessionToken() == "" {
 			return false
 		}
 
-		session, err := GetSession(webCon.SessionToken)
+		session, err := GetSession(webCon.GetSessionToken())
 		if err != nil {
 			l4g.Error(utils.T("api.websocket.invalid_session.error"), err.Error())
-			webCon.SessionToken = ""
-			webCon.SessionExpiresAt = 0
-			webCon.Session = nil
+			webCon.SetSessionToken("")
+			webCon.SetSession(model.Session{})
```

Comment:
Why change this? Setting the session to `nil` is more intuitive than setting it to an empty session. And you wouldn't have to complicate your conditionals by adding things like `|| len(currentSession.Token) == 0`.

Reply:
From the docs `Store(nil)` will panic

Reply:
Storing an _untyped_ nil will panic. Storing a typed nil is fine as it satisfies the constraint that "all calls to Store for a given Value must use values of the same concrete type." 

```go
package main

import (
	"sync/atomic"
)

type Foo struct{}

func main() {
	var x atomic.Value
	var foo *Foo
	x.Store(foo) // fine
	x.Store(nil) // panics
}
```

But even if storing a typed nil caused a panic, I would argue that you should work around rather than implementing a leaky abstraction. :P

Reply:
Might be somewhat unintuitive but we already have this abstraction in other places.  For example the context https://github.com/mattermost/platform/blob/master/api/context.go#L24 assumes an empty 
 unset session.

Reply:
If that was the real reason you changed it, you probably should have said that to begin with. I think a consistency argument here is a very big stretch, and almost definitely wasn't an actual consideration when you wrote this PR.

I'm not gonna block the PR on this request though. Gotta pick my battles. But I am going to be blunt: This is bad programming. And any precedence we may have set for this is a unacceptable excuse, even if you claim it's for consistency.

If you honestly and truly think this is better, go ahead and merge it. If you don't actually think this is better, but you're resisting the change for whatever reason (maybe you're away from your computer or too busy?), please let me know and I'll make the change.

Reply:
Okay I probably should have just dropped a 5/5 instead of all that (Sorry if it comes off as _too_ blunt.). But I'm still gonna leave it ultimately up to you.

Reply:
Sorry, I'm not trying to pick a fight here and you're right that's not the reason I changed it, but after looking at the code all I was saying is we already have this case in the workflow.  Meaning even if I make your suggested change I still think it's a good idea to leave in the extra check for `|| len(currentSession.Token) == 0` as defensive.

The React Native apps starts the socket without a session token so `NewWebConn` is called with `Session{}` and the real session is attached later by sending a websocket event.

Reply:
I have nothing against the defensive check. That's fine. And if you want the invalidated session to be the zero-value `model.Session`, even that's fine.

But this is what doesn't make sense:

```go
func (c *WebConn) GetSession() *model.Session {...}
func (c *WebConn) SetSession(v model.Session) {...}
```

If you were to comment those, the comment for `GetSession` would probably include something like...

> GetSession returns nil if and only if SetSession has never been invoked.

There's no reason for that behavior, misleading signature, or burden on the caller. If the you want the default session to be a zero-value session, just make `GetSession` return a value.